### PR TITLE
CustomCluster configuration memory optimization

### PR DESCRIPTION
**Pull request by DiazAlan**
Commit Message: Saves memory in cluster info
Additional Description: In cases where CustomClusterType is not used, the unique_ptr will be null and save memory
Risk Level: Low
Testing: Changed mock for new typing
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

File: envoy/upstream/upstream.h [hunk 1]
```diff
@@ -968,8 +968,7 @@ class ClusterInfo : public Http::FilterChainFactory {
   /**
    * @return the type of cluster, only used for custom discovery types.
    */
-  virtual const absl::optional<envoy::config::cluster::v3::Cluster::CustomClusterType>&
-  clusterType() const PURE;
+  virtual OptRef<envoy::config::cluster::v3::Cluster::CustomClusterType> clusterType() const PURE;
```
Comment (adisuissa) [inline envoy/upstream/upstream.h:971]: nit: return a const CustomClusterType

File: source/common/upstream/upstream_impl.h [hunk 2]
```diff
@@ -761,9 +761,12 @@ class ClusterInfoImpl : public ClusterInfo,
   extensionProtocolOptions(const std::string& name) const override;
   LoadBalancerType lbType() const override { return lb_type_; }
   envoy::config::cluster::v3::Cluster::DiscoveryType type() const override { return type_; }
-  const absl::optional<envoy::config::cluster::v3::Cluster::CustomClusterType>&
-  clusterType() const override {
-    return cluster_type_;
+
+  OptRef<envoy::config::cluster::v3::Cluster::CustomClusterType> clusterType() const override {
```
Comment (adisuissa) [inline source/common/upstream/upstream_impl.h:765]: nit: return const CustomClusterType.

**Review by adisuissa — COMMENTED** [review]
LGTM modulo nit.
Thanks!

File: source/common/upstream/upstream_impl.h (hunk 2 above)
Comment (DiazAlan) [reply source/common/upstream/upstream_impl.h]: Done

File: envoy/upstream/upstream.h (hunk 1 above)
Comment (DiazAlan) [reply envoy/upstream/upstream.h]: Done

**Comment by DiazAlan** [comment]
@yurykats @adisuissa

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!

/assign-from @envoyproxy/senior-maintainers

**Comment by adisuissa** [comment]
CI failure seems relevant to the PR changes. PTAL.

**Review by zuercher — DISMISSED** [review]

File: test/extensions/filters/network/redis_proxy/conn_pool_impl_test.cc [hunk 3]
```diff
@@ -1032,11 +1032,13 @@ TEST_F(RedisConnPoolImplTest, HostsAddedAndEndWithClusterRemoval) {
...
-  cluster_type.emplace();
-  cluster_type->set_name("envoy.clusters.redis");
+  envoy::config::cluster::v3::Cluster::CustomClusterType cluster_type_;
+  cluster_type_.set_name("envoy.clusters.redis");
+
+  OptRef<const envoy::config::cluster::v3::Cluster::CustomClusterType> cluster_type =
+      makeOptRef<const envoy::config::cluster::v3::Cluster::CustomClusterType>(cluster_type_);
   EXPECT_CALL(*cm_.thread_local_cluster_.cluster_.info_, clusterType())
-      .WillOnce(ReturnRef(cluster_type));
+      .WillOnce(Return(cluster_type));
```
Comment (adisuissa) [inline test/extensions/filters/network/redis_proxy/conn_pool_impl_test.cc:1041]: nit: Can this directly return `makeOptRef(cluster_type_)` (avoiding the temp var above)?

File: test/extensions/filters/network/redis_proxy/conn_pool_impl_test.cc [hunk 4]
```diff
@@ -1052,11 +1054,13 @@ TEST_F(RedisConnPoolImplTest, MakeRequestToRedisCluster) {
...
```
Comment (adisuissa) [inline test/extensions/filters/network/redis_proxy/conn_pool_impl_test.cc:1063]: Same as above

**Review by adisuissa — COMMENTED** [review]
Thanks, overall LGTM.
Left one minor comment/question.

File: test/extensions/filters/network/redis_proxy/conn_pool_impl_test.cc (hunk 3 above)
Comment (DiazAlan) [reply test/extensions/filters/network/redis_proxy/conn_pool_impl_test.cc]: Done

File: test/extensions/filters/network/redis_proxy/conn_pool_impl_test.cc (hunk 4 above)
Comment (DiazAlan) [reply test/extensions/filters/network/redis_proxy/conn_pool_impl_test.cc]: Done

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!
/assign-from @envoyproxy/senior-maintainers

**Review by zuercher — APPROVED** [review]

**Pull request merged**